### ISSUE TITLE
Can't get a simple XOR problem network to work, answer always array([0])

### ISSUE THREAD

**Issue by logloglog** [fchollet/keras#112]
I am trying to implement a XOR-problem solving network and can't seem to get it to work. Here's my code:
```
model = Sequential()
model.add(Dense(2,2))
model.add(Activation('sigmoid'))
model.add(Dense(2,1))
model.add(Activation('softmax'))
X = numpy.array([[0,0],[0,1],[1,0],[1,1]])
y = numpy.array([[0],[1],[1],[0]])
model.compile(loss='categorical_crossentropy', optimizer='sgd')
model.fit(X, y, nb_epoch=5, batch_size=32)
```
No matter what input I try, the answer to `predict_classes` is always `array([0])`, with `predict_proba` result always being `array([[ 1.]])`.

I have tried other setups, with tanh activation, `loss='mean_absolute_error'`, `optimizer='rmsprop'`, `nb_epoch=20, batch_size=16`, but there was no difference.

**Comment by fchollet** [comment]
XOR hasn't been an interesting Hello World for NNs since the 70s. Look at MNIST maybe. 

But if you really do want to get your code to work, notice that a batch_size of 32 over 4 samples means you are really doing one update per epoch. So 5 updates in total (5 epochs). Consider increasing that until it works.

**Comment by logloglog** [comment]
I'm not looking for an interesting example, I'm trying to learn how to use the library.
I have tried it with 400 and 4000 epochs, with `binary_crossentropy`. Doesn't change anything.

**Comment by fchollet** [comment]
A 3rd issue is the use of softmax. Do you know what softmax does? https://en.wikipedia.org/wiki/Softmax_function

**Comment by logloglog** [comment]
I don't think softmax is relevant here. I have tried constructing a net without it - again, no difference.

**Comment by fchollet** [comment]
Another thing that you might be doing wrong: unless you're using an Embedding layer (indexes) your data should be float. Try converting the input to float32.

**Comment by logloglog** [comment]
Tried using `numpy.array`'s `astype('float32')` everywhere. No difference.

**Comment by pranv** [comment]
I don't see you having an optimizer and then compiling.

**Comment by logloglog** [comment]
Did you read the code I posted?
```
model.compile(loss='categorical_crossentropy', optimizer='sgd')
```

**Comment by fchollet** [comment]
I don't have the time to get your code working, but in any case solving XOR with Keras sounds like using a Merlin rocket engine to make a grilled cheese sandwich. 

If you are genuinely interested in learning how to use Keras, I recommend you check out the [examples page](http://keras.io/examples/), the [examples folder](https://github.com/fchollet/keras/tree/master/examples) (6 scripts solving real problems), and the intro ["30 seconds to Keras"](http://keras.io/#getting-started-30-seconds-to-keras). 

If you are genuinely seeking a XOR operator, I recommend you use the `^` Python operator.

**Comment by logloglog** [comment]
If I weren't interested in learning how to use Keras, I wouldn't have raised this issue. If you don't have the time to help, then you shouldn't spend it writing a passive-aggressive retort either.

**Comment by nagadomi** [comment]
XOR example:
```
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
import numpy as np

np.random.seed(100)

model = Sequential()
model.add(Dense(2, 8))
model.add(Activation('relu'))
model.add(Dense(8, 1))
model.add(Activation('sigmoid'))
X = np.array([[0,0],[0,1],[1,0],[1,1]], "float32")
y = np.array([[0],[1],[1],[0]], "float32")
sgd = SGD(lr=0.1)
model.compile(loss='binary_crossentropy', optimizer=sgd)
model.fit(X, y, nb_epoch=1000, batch_size=1)

print(model.predict_proba(X))
```
```
[[ 0.00262257]
 [ 0.99859906]
 [ 0.9986052 ]
 [ 0.00181695]]
```

**Comment by Ori226** [comment]
I tried to run this example without success. When I change the activation from 'relu' to 'tanh' it worked better:


from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
import numpy as np

np.random.seed(100)

model = Sequential()
model.add(Dense(2, 8))
model.add(Activation('tanh'))
model.add(Dense(8, 1))
model.add(Activation('sigmoid'))
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])
sgd = SGD(lr=0.1)
model.compile(loss='binary_crossentropy', optimizer=sgd)

**Comment by c0d3rman** [comment]
I was trying to make my own XOR net and experienced similar result. In the end what worked for me was looking at @nagadomi's code and using ReLu instead of sigmoid for the activation on the input layer. Using tanh as per @Ori226's suggestion worked as well, but using sigmoid just makes the predicted values be randomly distributed very close to 0.5. This goes away when I removed `class_mode='binary'` from my `model.compile`. Anyone know why?

**Comment by lukedeo** [comment]
I really think looking at the `examples` folder would be a better use of your time. The default parameters on solvers in Keras are meant to deal with real-world scale problems...

**Comment by outlace** [comment]
Empirically it seems a network of only ReLU units cannot learn the simple XOR problem. Leaky ReLU units, however, can learn it quite nicely.

**Comment by dcgrigsby** [comment]
Here's an XOR net that works with the new API:

```python
import numpy as np
from keras.models import Sequential
from keras.layers.core import Activation, Dense
from keras.optimizers import SGD

X = np.zeros((4, 2), dtype='uint8')
y = np.zeros(4, dtype='uint8')

X[0] = [0, 0]
y[0] = 0
X[1] = [0, 1]
y[1] = 1
X[2] = [1, 0]
y[2] = 1
X[3] = [1, 1]
y[3] = 0

model = Sequential()
model.add(Dense(2, input_dim=2))
model.add(Activation('sigmoid'))
model.add(Dense(1))
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='mean_squared_error', optimizer=sgd, class_mode="binary")

history = model.fit(X, y, nb_epoch=10000, batch_size=4, show_accuracy=True, verbose=0)

print model.predict(X)
```

**Comment by amit4111989** [comment]
For anyone who cares, this works perfectly. Gives me a loss of 0.0010 without hitting convergence. 
The convergence will hit after 3500+ epochs, so don't bother running it for so long. 

The takeaway from this, 

- updating 4 times in one epoch (total updates = 4*2000 = 8000 updates) works! 
- combination of sigmoid acitvations is perfect for this example.
- decay=1e-6, momentum=0.9, nesterov=True in the SGD operator makes all the difference

```

from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
import numpy as np

np.random.seed(100)

model = Sequential()
model.add(Dense(2, input_dim=2))
model.add(Activation('sigmoid'))
model.add(Dense(1, input_dim=2))
model.add(Activation('sigmoid'))
X = np.array([[0,0],[0,1],[1,0],[1,1]], "float32")
y = np.array([[0],[1],[1],[0]], "float32")
sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', optimizer=sgd)
model.fit(X, y, nb_epoch=2000, batch_size=1)

print(model.predict(X))

```

**Comment by Miail** [comment]
I just tried your xor example I am getting an loss of 0.3 and it even predicts one of the cases incorrect.

**Comment by Miail** [comment]
This worked for me

```
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import Adam
import numpy as np

np.random.seed(100)

model = Sequential()
model.add(Dense(units = 3, input_dim=2, activation = 'tanh'))
model.add(Dense(units = 1, activation = 'sigmoid'))
X = np.array([[0,0],[0,1],[1,0],[1,1]], "float32")
y = np.array([[0],[1],[1],[0]], "float32")
adam = Adam(lr=0.1)
model.compile(loss='binary_crossentropy', optimizer=adam)
model.fit(X, y, nb_epoch=200, batch_size=4,verbose=1)

print(model.predict_classes(X))

```

**Comment by czcz1024** [comment]
i try your code,and i have a question
why you layer1 units choose 3,i try change it to 2,and the result sometimes wrong,sometimes right
i don't know why

**Comment by ibrahiminfinite** [comment]
Here is my implementation of XOR with sigmoid activation, logcosh loss and Adam optimizer

```
from keras.models import Sequential
from keras.layers import Dense, Activation
import numpy as np

X = np.array([[0,0],[0,1],[1,0],[1,1]])
Y = np.array([[0], [1], [1], [0]])

model = Sequential()
model.add(Dense(2, input_dim=2))
model.add(Activation('sigmoid'))
model.add(Dense(1))
model.add(Activation('sigmoid'))

model.compile(loss='logcosh', optimizer='Adam', metrics=['accuracy'])
model.fit(X, Y, epochs=16000)

print(model.predict(X))
```

Output :
```
[[0.00858523]
 [0.98780406]
 [0.9877358 ]
 [0.01478751]]

```

**Comment by jeffheaton** [comment]
Such hostility...  XOR is still a great for certain instances.  I use XOR as an example in a class to show my students how I can extract the weights from Keras and calculate the XOR output on the whiteboard.  I do not have a big enough whiteboard for MINST.  

I have several examples of XOR and how to extract/calculate the weights on a very small (2-hidden layer network).

https://github.com/jeffheaton/t81_558_deep_learning/blob/master/t81_558_class03_tensor_flow.ipynb

**Comment by j-rachit** [comment]
@amit4111989 
> * decay=1e-6, momentum=0.9, nesterov=True in the SGD operator makes all the difference

> sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)

Why do those particular parameters work, but not the default ones that are supplied by the Keras API? What changes between them?
